runtime/parachains/src/hrmp.rs

```diff
@@ -2008,3 +2015,244 @@ mod tests {
+		let recipient_origin: crate::Origin = to.into();
+
+		let output = [(sender, sender_origin.clone()), (recipient, recipient_origin.clone())];
+
+		// Make both a parachain if they are already not.
+		if !Paras::<T>::is_parachain(sender) {
+			register_parachain_with_balance::<T>(sender, deposit);
+		}
+		if !Paras::<T>::is_parachain(recipient) {
+			register_parachain_with_balance::<T>(recipient, deposit);
+		}
+
+		assert_ok!(Hrmp::<T>::hrmp_init_open_channel(
+			sender_origin.clone().into(),
+			recipient,
+			capacity,
+			message_size
+		));
+		if matches!(until, ParachainSetupStep::Requested) {
+			return output;
+		}
+
+		assert_ok!(Hrmp::<T>::hrmp_accept_open_channel(recipient_origin.into(), sender));
+		if matches!(until, ParachainSetupStep::Accepted) {
+			return output;
+		}
+
+		Hrmp::<T>::process_hrmp_open_channel_requests(&Configuration::<T>::config());
+		if matches!(until, ParachainSetupStep::Established) {
+			return output;
+		}
+
+		let channel_id = HrmpChannelId { sender, recipient };
+		assert_ok!(Hrmp::<T>::hrmp_close_channel(sender_origin.clone().into(), channel_id));
+		if matches!(until, ParachainSetupStep::CloseRequested) {
+			// NOTE: this is just for expressiveness, otherwise the if-statement could be omitted.
+			return output;
+		}
+
+		output
+	}
+
+	frame_benchmarking::benchmarks! {
+		where_clause { where <T as frame_system::Config>::Origin: From<crate::Origin> }
+
+		hrmp_init_open_channel {
+			let sender_id: ParaId = 1u32.into();
+			let sender_origin: crate::Origin = 1u32.into();
+
+			let recipient_id: ParaId = 2u32.into();
+
+			// make sure para is registered, and has enough balance.
+			let deposit: BalanceOf<T> = Configuration::<T>::config().hrmp_sender_deposit.unique_saturated_into();
+			register_parachain_with_balance::<T>(sender_id, deposit);
+			register_parachain_with_balance::<T>(recipient_id, deposit);
+
+			let capacity = Configuration::<T>::config().hrmp_channel_max_capacity;
+			let message_size = Configuration::<T>::config().hrmp_channel_max_message_size;
+		}: _(sender_origin, recipient_id, capacity, message_size)
+		verify {
+			assert_last_event::<T>(
+				Event::<T>::OpenChannelRequested(sender_id, recipient_id, capacity, message_size).into()
+			);
+		}
+
+		hrmp_accept_open_channel {
+			let [(sender, _), (recipient, recipient_origin)] =establish_para_connection::<T>(1, 2, ParachainSetupStep::Requested);
+		}: _(recipient_origin, sender)
+		verify {
+			assert_last_event::<T>(Event::<T>::OpenChannelAccepted(sender, recipient).into());
+		}
+
+		hrmp_close_channel {
+			let [(sender, sender_origin), (recipient, _)] = establish_para_connection::<T>(1, 2, ParachainSetupStep::Established);
+			let channel_id = HrmpChannelId { sender, recipient };
+		}: _(sender_origin, channel_id.clone())
+		verify {
+			assert_last_event::<T>(Event::<T>::ChannelClosed(sender, channel_id).into());
+		}
+
+		force_clean_hrmp {
+			// TODO: reasonable constant/config for these.
```

Comment:
Proper way to solve this, for my future self: These values come from top level `struct HostConfiguration` and it seems like all of the values are upgradable in the runtime through root. So we can't really make any assumption about them.

We need to put a bound in the upper bound of channel counts in the `configurations` pallet, and never allow even governance to go beyond that. Then we use that bound here as well.

Reply:
yup, we need this :)